File: p189.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        k %= nums.size();
        vector<int> rotated;
        for(int i = nums.size() - k; i < nums.size(); ++i) {
            rotated.push_back(nums[i]);
        }
        for(int j = 0; j < nums.size() - k; ++j) {
            rotated.push_back(nums[j]);
        }
        nums = rotated;
    }
};

File: p189.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        def gcd(a, b):
            if a%b == 0:
                return b
            return gcd(b, a%b)

        n = len(nums)
        k %= n
        if k == 0:
            return
        cycles = gcd(n, k)
        for idx in range(cycles):
            prev = nums[idx-k]
            for _ in range(n//cycles):
                tmp = nums[idx]
                nums[idx] = prev
                prev = tmp
                idx = (idx+k)%n
